File: medium/1372. Longest ZigZag Path in a Binary Tree/1372. Longest ZigZag Path in a Binary Tree.go
package main

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Direction int

const (
	Right Direction = 1
	Left  Direction = -1
)

func dfs(root *TreeNode, dir Direction, sum int) int {
	if root == nil {
		return sum
	}

	if root.Left == nil && root.Right == nil {
		return sum + 1
	}

	if dir == Right {
		return max(dfs(root.Left, Left, sum+1), dfs(root.Right, Right, 0))
	} else {
		return max(dfs(root.Right, Right, sum+1), dfs(root.Left, Left, 0))
	}
}

func longestZigZag(root *TreeNode) int {
	if root == nil {
		return 0
	}

	return max(dfs(root.Left, Left, 0), dfs(root.Right, Right, 0))
}

File: medium/1372. Longest ZigZag Path in a Binary Tree/1372. Longest ZigZag Path in a Binary Tree.py
from typing import Literal, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        # 1 is the right direction
        # -1 is the left direction
        def dfs(node: Optional[TreeNode], path: int, direction: Literal[1, -1]) -> int:
            if not node:
                return path

            if not node.left and not node.right:
                return path + 1

            if direction == 1:
                return max(dfs(node.left, path + 1, -1), dfs(node.right, 0, 1))
            else:
                return max(dfs(node.right, path + 1, 1), dfs(node.left, 0, -1))

        return max(dfs(root.left, 0, -1), dfs(root.right, 0, 1))
